jgrapht-core/src/main/java/org/jgrapht/alg/flow/min_cost/CapacityScalingMinimumCostFlow.java

```diff
@@ -0,0 +1,892 @@
+     *
+     * @param start the start of the augmenting path
+     * @param end   the end of the augmenting path
+     */
+    private void augmentPath(Node start, Node end) {
+        // compute delta to augment
+        int valueToAugment = Math.min(start.excess, -end.excess);
+        for (Arc arc = end.parentArc; arc != null; arc = arc.revArc.head.parentArc) {
+            valueToAugment = Math.min(valueToAugment, arc.residualCapacity);
+        }
+        if (DEBUG) {
+            ArrayList<Node> stack = new ArrayList<>();
+            for (Arc arc = end.parentArc; arc != null; arc = arc.revArc.head.parentArc) {
+                stack.add(arc.head);
+            }
+            stack.add(start);
+            System.out.println("Printing augmenting path");
+            for (int i = stack.size() - 1; i > 0; i--) {
+                System.out.print(stack.get(i).id + " -> ");
+            }
+            System.out.println(stack.get(0).id + ", delta = " + valueToAugment);
+        }
+        // augmenting the flow
+        end.excess += valueToAugment;
+        for (Arc arc = end.parentArc; arc != null; arc = arc.revArc.head.parentArc) {
+            arc.sendFlow(valueToAugment);
+        }
+        start.excess -= valueToAugment;
+    }
+
+    /**
+     * Finishes the computation by checking the flow feasibility, computing arc flows, and creating an instance
+     * of {@link MinimumCostFLow}. The resulting flow mapping contains all edges of the specified minimum cost
+     * flow problem.
+     *
+     * @return the solution to the minimum cost flow problem
+     */
+    private MinimumCostFLow<E> finish() {
+        Map<E, Integer> flowMap = new HashMap<>(m);
+        double totalCost = 0;
+        // check feasibility
+        for (Arc arc = nodes[n].firstNonSaturated; arc != null; arc = arc.next) {
+            if (arc.revArc.residualCapacity > 0) {
+                throw new IllegalArgumentException("Specified flow network problem has no feasible solution");
+            }
+        }
+        // create the solution object
+        for (int i = 0; i < m; i++) {
+            E graphEdge = graphEdges.get(i);
+            Arc arc = arcs[i];
+            int flowOnArc = arc.revArc.residualCapacity; // this value equals to the flow on the initial arc
+            if (problem.graph.getEdgeWeight(graphEdge) < 0) {
+                // the initial arc goes in the opposite direction
+                flowOnArc = problem.upperCapacityMap.get(graphEdge) - problem.lowerCapacityMap.getOrDefault(graphEdge, 0) - flowOnArc;
+            }
+            flowOnArc += problem.lowerCapacityMap.getOrDefault(graphEdge, 0);
+            flowMap.put(graphEdge, flowOnArc);
+            totalCost += flowOnArc * problem.graph.getEdgeWeight(graphEdge);
+        }
+        return new MinimumCostFlowImpl<>(totalCost, flowMap);
+    }
+
+    /**
+     * Tests the optimality conditions after a flow of minimum cost has been computed.
```

Comment:
This method is confusing. Doesn't the implementation guarantee that the final solution is optimal? Why would testOptimality ever return false? Moreover, if it does return false, then what? Please clarify this in the documentation: Under what conditions would this method potentially return false?